File: 0.1/quadstream2010/src/main/webapp/js/rtree.js
dojo.declare("RTree", null, {
	constructor : function() {
		this.maxEntries = 5;
		this.temp = new Rectangle(0, 0, 0, 0);
		this.xHigh = new Rectangle(-Number.MAX_VALUE, 0, 0, 0);
		this.xLow = new Rectangle(0, 0, Number.MAX_VALUE, 0);
		this.yHigh = new Rectangle(0, -Number.MAX_VALUE, 0, 0);
		this.yLow = new Rectangle(0, 0, 0, Number.MAX_VALUE);
	},
	insert : function(e) {
		if (this.root == undefined) {
			this.root = new RTreeNode(true);
			this.root.addEntry(e);
		} else {
			var l = this.chooseLeaf(e, this.root);
			l.addEntry(e);
			var ll;
			if (l.entries.length > this.maxEntries)
				ll = this.splitNode(l);
			var nn = this.adjustTree(l, ll);
			if (nn != null) {
				var n = this.root;
				this.root = new RTreeNode(false);
				this.root.addEntry(n);
				this.root.addEntry(nn);
			}
		}
	},
	search : function(result, r) {
		if (this.root != null)
			this.innerSearch(result, r, this.root);
	},
	innerSearch : function(result, r, n) {
		for (i in n.entries) {
			var c = n.entries[i];
			if (c.intersects(r))
				if (n.isLeaf)
					result.push(c);
				else
					this.innerSearch(result, r, c);
		}
	},
	forEachSearch : function(r,callback) {
		if (this.root != null)
			this.innerforEachSearch(r,callback, this.root);
	},
	innerforEachSearch : function(r,callback, n) {
		for (i in n.entries) {
			var c = n.entries[i];
			if (c.intersects(r))
				if (n.isLeaf)
					callback(c);
				else
					this.innerforEachSearch(r,callback, c);
		}
	},
	adjustTree : function(n, nn) {
		if (n == this.root)
			return nn;
		var p = n.parent;
		p.adjust();
		var pp;
		if (nn != undefined) {
			p.addEntry(nn);
			if (p.entries.length > this.maxEntries)
				pp = this.splitNode(p);
		}
		return this.adjustTree(p, pp);
	},
	splitNode : function(l) {
		var group = l.entries;
		var ll = new RTreeNode(l.isLeaf);
		l.entries = null;
		this.linearPickSeeds(l, ll, group);
		var nodes = [ l, ll ];
		for ( var i = 0; i < group.length; i++) {
			var e = group[i];
			if (e != undefined)
				this.smallestChanged(e, nodes).addEntry(e);
		}
		return ll;
	},
	linearPickSeeds : function(l, ll, group) {
		var xMin = Number.MAX_VALUE;
		var yMin = Number.MAX_VALUE;
		var xMax = -Number.MAX_VALUE;
		var yMax = -Number.MAX_VALUE;

		var xHigh = this.xHigh;
		var xLow = this.xLow;
		var yHigh = this.yHigh;
		var yLow = this.yLow;

		for ( var i = 0; i < group.length; i++) {
			var r = group[i];
			if (r.x1 > xHigh.x1)
				xHigh = r;
			if (r.x2 < xLow.x2)
				xLow = r;
			if (r.y1 > yHigh.y1)
				yHigh = r;
			if (r.y2 < yLow.y2)
				yLow = r;
			if (r.x2 > xMax)
				xMax = r.x2;
			if (r.x1 < xMin)
				xMin = r.x1;
			if (r.y2 > yMax)
				yMax = r.y2;
			if (r.y1 < yMin)
				yMin = r.y1;
		}
		var w = xMax - xMin;
		var h = yMax - yMin;
		var xGap = xHigh.x1 - xLow.x2;
		var yGap = yHigh.y1 - yLow.y2;
		if (xGap / w > yGap / h) {
			l.addEntry(xLow);
			ll.addEntry(xHigh);

			for ( var i = 0; i < group.length; i++)
				if (group[i] == xLow || group[i] == xHigh)
					group[i] = undefined;
		} else {
			l.addEntry(yLow);
			ll.addEntry(yHigh);
			for ( var i = 0; i < group.length; i++)
				if (group[i] == yLow || group[i] == yHigh)
					group[i] = undefined;
		}
	},
	chooseLeaf : function(e, n) {
		return n.isLeaf ? n : this.chooseLeaf(e, this.smallestChanged(e,
				n.entries));
	},
	smallestChanged : function(e, entries) {
		var smallestChange = Number.MAX_VALUE;
		var f;
		for ( var i = 0; i < entries.length; i++) {
			o = entries[i];
			this.temp.set(o);
			this.temp.add(e);
			var change = this.temp.area() - o.area();
			if (change < smallestChange) {
				f = o;
				smallestChange = change;
			}
		}
		return f;
	}
});
dojo.declare("RTreeNode", Rectangle, {
	constructor : function(isLeaf) {
		this.isLeaf = isLeaf;
		this.isRTreeNode = true;
	},
	addEntry : function(e) {
		if (this.entries == undefined) {
			this.entries = [];
			this.set(e);
		} else
			this.add(e);
		this.entries.push(e);
		if (e.isRTreeNode)
			e.parent = this;
	},
	adjust : function() {
		var first = true;
		for ( var i = 0; i < this.entries.length; i++) {
			var n = this.entries[i];
			if (first) {
				this.set(n);
				first = false;
			} else
				this.add(n);
		}
	}
});

File: 0.1/quadstream2010/Quadstream02/src/rtree/RTree.java
package rtree;

import geometry.Rectangle;

import java.util.ArrayList;
import java.util.List;

public class RTree {
	RTreeNode root;
	Rectangle temp = new Rectangle();
	public static int maxEntries = 2;

	public void insert(Rectangle e) {
		if (root == null) {
			root = new RTreeNode(true);
			root.addEntry(e);
		} else {
			RTreeNode l = chooseLeaf(e, root);
			l.addEntry(e);
			RTreeNode ll = null;
			if (l.entries.size() > maxEntries)
				ll = splitNode(l);
			RTreeNode nn = adjustTree(l, ll);
			if (nn != null) {
				RTreeNode n = root;
				root = new RTreeNode(false);
				root.addEntry(n);
				root.addEntry(nn);
			}
		}
	}

	private RTreeNode adjustTree(RTreeNode n, RTreeNode nn) {
		if (n == root)
			return nn;
		RTreeNode p = n.parent;
		p.adjust();
		RTreeNode pp = null;
		if (nn != null) {
			p.addEntry(nn);
			if (p.entries.size() > maxEntries)
				pp = splitNode(p);
		}
		return adjustTree(p, pp);
	}

	/**
	 * L should be an overfull leaf node. This function partitions the entries
	 * of l between l and the returned node.
	 */
	private RTreeNode splitNode(RTreeNode l) {
		List<Rectangle> group = l.entries;
		RTreeNode ll = new RTreeNode(l.isLeaf);
		l.entries = null;

		linearPickSeeds(l, ll, group);
		List<Rectangle> nodes = new ArrayList<Rectangle>(2);
		nodes.add(l);
		nodes.add(ll);
		for (Rectangle e : group)
			((RTreeNode) smallestChanged(e, nodes)).addEntry(e);
		return ll;
	}

	private void linearPickSeeds(RTreeNode l, RTreeNode ll, List<Rectangle> group) {
		double xMin = Double.MAX_VALUE;
		double yMin = Double.MAX_VALUE;
		double xMax = -Double.MAX_VALUE;
		double yMax = -Double.MAX_VALUE;

		Rectangle xHigh = new Rectangle(-Double.MAX_VALUE, 0, 0, 0);
		Rectangle xLow = new Rectangle(Double.MAX_VALUE, 0, 0, 0);
		Rectangle yHigh = new Rectangle(0, -Double.MAX_VALUE, 0, 0);
		Rectangle yLow = new Rectangle(0, Double.MAX_VALUE, 0, 0);

		for (Rectangle r : group) {
			if (r.x > xHigh.x)
				xHigh = r;
			if (r.x + r.width < xLow.x + xLow.width)
				xLow = r;
			if (r.y > yHigh.y)
				yHigh = r;
			if (r.y + r.height < yLow.y + xLow.height)
				yLow = r;
			if (r.x + r.width > xMax)
				xMax = r.x + r.width;
			if (r.x < xMin)
				xMin = r.x;
			if (r.y + r.height > yMax)
				yMax = r.y + r.height;
			if (r.y < yMin)
				yMin = r.y;
		}
		double w = xMax - xMin;
		double h = yMax - yMin;
		double xGap = xHigh.x - (xLow.x + xLow.width);
		double yGap = yHigh.y - (yLow.y + yLow.height);
		if (xGap / w > yGap / h) {
			group.remove(xLow);
			l.addEntry(xLow);
			group.remove(xHigh);
			ll.addEntry(xHigh);
		} else {
			group.remove(yLow);
			l.addEntry(yLow);
			group.remove(yHigh);
			ll.addEntry(yHigh);
		}
	}

	private RTreeNode chooseLeaf(Rectangle e, RTreeNode n) {
		if (n.isLeaf())
			return n;
		else {
			Rectangle f = smallestChanged(e, n.entries);
			return chooseLeaf(e, (RTreeNode) f);
		}
	}

	/**
	 * Returns the node from nodes whose area would change the least if e were
	 * added to it.
	 */
	private Rectangle smallestChanged(Rectangle e, List<Rectangle> entries) {
		double smallestChange = Double.MAX_VALUE;
		Rectangle f = null;
		for (Rectangle i : entries) {
			double originalArea = i.width * i.height;
			temp.setRect(i);
			temp.add(e);
			double updatedArea = temp.width * temp.height;
			double change = updatedArea - originalArea;
			if (change < smallestChange) {
				f = i;
				smallestChange = change;
			}
		}
		return f;
	}

	public void unmarkAll() {
		unmarkAll(root);
	}

	private void unmarkAll(RTreeNode n) {
		if (n.entries != null)
			for (Rectangle e : n.entries)
				if (e instanceof RTreeNode) {
					RTreeNode r = (RTreeNode) e;
					r.visited = false;
					unmarkAll(r);
				}
	}

	public void search(List<Rectangle> result, Rectangle r) {
		if (root != null)
			innerSearch(result, r, root);
	}

	private void innerSearch(List<Rectangle> result, Rectangle r, RTreeNode n) {
		n.visited = true;
		for (Rectangle c : n.entries)
			if (c.intersects(r))
				if (n.isLeaf)
					result.add(c);
				else
					innerSearch(result, r, (RTreeNode) c);
	}

}
